CovertInfiltration/Src/CovertInfiltration/Classes/X2StrategyElement_InfiltrationRewards.uc

```diff
@@ -764,3 +770,107 @@ static function array<XComGameState_Tech> GetCandidatesForTechRush()
+	RewardState.Quantity = Amount;
+}
+
+static function GiveMultiPOIReward(XComGameState NewGameState, XComGameState_Reward RewardState, optional StateObjectReference AuxRef, optional bool bOrder = false, optional int OrderHours = -1)
+{
+	local XComGameStateHistory History;
+	local XComGameState_PointOfInterest POIState;
+	local XComGameState_HeadquartersResistance ResHQ;
+	local int idx;
+
+	History = `XCOMHISTORY;
+	ResHQ = class'UIUtilities_Strategy'.static.GetResistanceHQ();
+
+	for (idx = 0; idx < RewardState.Quantity; idx++)
+	{
+		POIState = XComGameState_PointOfInterest(History.GetGameStateForObjectID(ResHQ.ChoosePOI(NewGameState).ObjectID));
+
+		if (POIState != none)
+		{
+			POIState = XComGameState_PointOfInterest(NewGameState.ModifyStateObject(class'XComGameState_PointOfInterest', POIState.ObjectID));
+			POIState.Spawn(NewGameState);
+		}
+	}
+}
+
+static function GenerateMultiPOIReward (XComGameState_Reward RewardState, XComGameState NewGameState, optional float RewardScalar = 1.0, optional StateObjectReference AuxRef)
+{
+	RewardState.SetReward(, GetMultiPOIReward());
+}
+
+static function int GetMultiPOIReward()
+{
+	local int SitesMin;
+	local int SitesMax;
+	
+	SitesMin = `ScaleStrategyArrayInt(default.ScanSitesRewardMin);
+	SitesMax = `ScaleStrategyArrayInt(default.ScanSitesRewardMax);
+	
+	return SitesMin + `SYNC_RAND_STATIC(SitesMax - SitesMin);
+}
+
+static function X2DataTemplate CreateBlackMarketRewardTemplate()
+{
+	local X2RewardTemplate Template;
+
+	`CREATE_X2Reward_TEMPLATE(Template, 'Reward_BlackMarket');
+
+	Template.SetRewardFn = SetBlackMarketReward;
+	Template.GiveRewardFn = GiveBlackMarketReward;
+	Template.GetRewardIconFn = class'X2StrategyElement_DefaultRewards'.static.GetGenericRewardIcon;
+	Template.IsRewardAvailableFn = IsBlackMarketRewardAvailable;
+
+	return Template;
+}
+
+static function SetBlackMarketReward(XComGameState_Reward RewardState, optional StateObjectReference RewardObjectRef, optional int Amount)
+{
+	RewardState.RewardObjectReference = RewardObjectRef;
+	RewardState.Quantity = Amount;
+}
+
+static function GiveBlackMarketReward(XComGameState NewGameState, XComGameState_Reward RewardState, optional StateObjectReference AuxRef, optional bool bOrder = false, optional int OrderHours = -1)
+{
+	local XComGameState_BlackMarket BlackMarketState;
+
+	BlackMarketState = XComGameState_BlackMarket(`XCOMHISTORY.GetSingleGameStateObjectForClass(class'XComGameState_BlackMarket'));
+	BlackMarketState = XComGameState_BlackMarket(NewGameState.ModifyStateObject(class'XComGameState_BlackMarket', BlackMarketState.ObjectID));
+
+	BlackMarketState.bNeedsScan = true;
+	BlackMarketState.bForceClosed = false;
+	BlackMarketState.bNeedsAppearedPopup = true;
+	BlackMarketState.SetScanHoursRemaining(`ScaleStrategyArrayInt(BlackMarketState.default.MinScanDays), `ScaleStrategyArrayInt(BlackMarketState.default.MaxScanDays));
+	BlackMarketState.SetContinent();
+	BlackMarketState.ResetBlackMarketGoods(NewGameState);
```

Comment:
Or you can just do

```suggestion
	BlackMarketState.ShowBlackMarket(NewGameState, true);
```

Reply:
??? you told me to manually replicate ShowBlackMarket instead of calling it, I don't understand

Reply:
Back then I missed the fact that you can pass `true` as second argument to force it to show :(